File: translator.py
import cv2
import numpy as np
import json
import argparse
import os

def frame_to_ascii(frame, ascii_cols, character_set):

    gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # account for non-square pixels in terminals
    height, width = gray_frame.shape
    cell_width = width / ascii_cols
    cell_height = cell_width * 2
    aspect_ratio_corrected_rows = int(height / cell_height)

    resized_gray = cv2.resize(gray_frame, (ascii_cols, aspect_ratio_corrected_rows))

    ascii_frame = []
    for r in range(aspect_ratio_corrected_rows):
        row_of_chars = []
        for c in range(ascii_cols):
            pixel_brightness = resized_gray[r, c]
            char_index = int(pixel_brightness / 256 * len(character_set))
            row_of_chars.append(character_set[char_index])
        ascii_frame.append("".join(row_of_chars))
    return "\n".join(ascii_frame)

def translate_video(video_path, output_file, ascii_width=120):

    if not os.path.exists(video_path):
        print("Error: Video file not found at '{}'".format(video_path))
        return

    video_capture = cv2.VideoCapture(video_path)

    if not video_capture.isOpened():
        print("Error: Could not open video file '{}'. Check path and format.".format(video_path))
        return

    fps = video_capture.get(cv2.CAP_PROP_FPS)
    total_frames = int(video_capture.get(cv2.CAP_PROP_FRAME_COUNT))
    print('Processing video: {}'.format(video_path))
    print('FPS: {}'.format(fps))
    print('Total frames: {}'.format(total_frames))
    print('ASCII width: {} columns'.format(ascii_width))

    list_of_ascii_frames = []
    frames_processed = 0

    # this is a pretty standard character set for this kind of thing
    CHARACTER_SET = " .:-=+*#%@"

    while True:
        was_successful, frame = video_capture.read()
        if not was_successful:
            break

        ascii_output = frame_to_ascii(frame, ascii_width, CHARACTER_SET)
        list_of_ascii_frames.append(ascii_output)

        frames_processed += 1
        if frames_processed % 100 == 0:
            print("Processed {}/{} frames...".format(frames_processed, total_frames))

    video_capture.release()
    print("Processed all {} frames.".format(frames_processed))

    output_data = {
        "fps": fps,
        "width": ascii_width,
        "height": len(list_of_ascii_frames[0].split('\n')) if list_of_ascii_frames else 0,
        "frames": list_of_ascii_frames
    }

    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(output_data, f, ensure_ascii=False, indent=2)

    print("ASCII video data saved to '{}'".format(output_file))

if __name__ == "__main__":

    arg_parser = argparse.ArgumentParser(description="Convert MP4 videos to ASCII art .obj files.")

    arg_parser.add_argument("input_video", help="Path to the input MP4 video file.")
    arg_parser.add_argument("output_obj", help="Path to the output .obj file.")
    arg_parser.add_argument("--width", type=int, default=120, help="Width of the ASCII art in characters. Default: 120")

    parsed_args = arg_parser.parse_args()

    translate_video(parsed_args.input_video, parsed_args.output_obj, parsed_args.width)

File: translator.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <opencv2/opencv.hpp>

void f_to_a(cv::Mat& frame, int w, const char* cset, std::ostream& os) {
    cv::Mat gray;
    cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);

    int h = gray.rows;
    int width = gray.cols;
    float cw = (float)width / w;
    float ch = cw * 2;
    int rh = (int)(h / ch);

    cv::Mat res;
    cv::resize(gray, res, cv::Size(w, rh), 0, 0, cv::INTER_LINEAR);

    os << "";
    for (int r = 0; r < rh; r++) {
        for (int c = 0; c < w; c++) {
            unsigned char br = res.at<unsigned char>(r, c);
            int cidx = (int)(br / 256.0 * strlen(cset));
            os << cset[cidx];
        }
        if (r < rh - 1) {
            os << "\n";
        }
    }
    os << "";
}

int main(int argc, char** argv) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <input_video> <output_obj> [--width <w>]" << std::endl;
        return 1;
    }

    std::string vid_path = argv[1];
    std::string out_path = argv[2];
    int w = 120;

    for (int i = 3; i < argc; i++) {
        if (std::string(argv[i]) == "--width" && i + 1 < argc) {
            w = std::stoi(argv[i + 1]);
            break;
        }
    }

    cv::VideoCapture cap(vid_path);
    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open video file '" << vid_path << "'." << std::endl;
        return 1;
    }

    double fps = cap.get(cv::CAP_PROP_FPS);
    int total = (int)cap.get(cv::CAP_PROP_FRAME_COUNT);
    std::cout << "Processing video: " << vid_path << std::endl;
    std::cout << "FPS: " << fps << std::endl;
    std::cout << "Total frames: " << total << std::endl;
    std::cout << "ASCII width: " << w << " columns" << std::endl;

    std::ofstream f(out_path);
    if (!f.is_open()) {
        std::cerr << "Error: Could not open output file '" << out_path << "'." << std::endl;
        return 1;
    }

    const char* cset = " .:-=+*#%@";
    int processed = 0;
    bool first = true;

    cv::Mat frame;
    int rh = 0;

    if (cap.read(frame)) {
        float ch = ((float)frame.cols / w) * 2;
        rh = (int)(frame.rows / ch);
    }
    cap.set(cv::CAP_PROP_POS_FRAMES, 0);

    f << "{\n";
    f << "  \"fps\": " << fps << ",\n";
    f << "  \"width\": " << w << ",\n";
    f << "  \"height\": " << rh << ",\n";
    f << "  \"frames\": [\n    ";

    while (cap.read(frame)) {
        if (!first) {
            f << ",\n    ";
        }
        f_to_a(frame, w, cset, f);
        first = false;

        processed++;
        if (processed % 100 == 0) {
            std::cout << "Processed " << processed << "/" << total << " frames..." << std::endl;
        }
    }

    f << "\n  ]\n}\n";

    std::cout << "Processed all " << processed << " frames." << std::endl;
    std::cout << "ASCII video data saved to '" << out_path << "'" << std::endl;

    return 0;
}